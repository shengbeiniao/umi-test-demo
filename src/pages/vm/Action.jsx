import { PureComponent } from 'react'
import { connect } from 'dva'
import { Button, Dropdown, Menu, Icon } from 'antd'
import style from './style.less'

@connect()
export default class extends PureComponent {
  doAction = action => {
    const { current, dispatch } = this.props
    dispatch({
      type: 'vm/doAction',
      payload: {
        current,
        action
      }
    })
  }

  isEnable = type => {
    const { current } = this.props
    const { state } = current
    switch (type) {
      case 'start':
        return ['已停止'].indexOf(state) >= 0
      case 'stop':
        return ['运行中', '已暂停'].indexOf(state) >= 0
      case 'resume':
        return ['已暂停'].indexOf(state) >= 0
      case 'pause':
        return ['运行中'].indexOf(state) >= 0
    }
  }

  render() {
    const { current } = this.props

    const menu = (
      <Menu data-testid="menu">
        <Menu.Item
          data-testid="start"
          onClick={() => this.doAction('start')}
          disabled={!this.isEnable('start')}>
          启动
        </Menu.Item>
        <Menu.Item
          data-testid="stop"
          onClick={() => this.doAction('stop')}
          disabled={!this.isEnable('stop')}>
          停止
        </Menu.Item>
        <Menu.Item
          data-testid="pause"
          onClick={() => this.doAction('pause')}
          disabled={!this.isEnable('pause')}>
          暂停
        </Menu.Item>
        <Menu.Item
          data-testid="resume"
          onClick={() => this.doAction('resume')}
          disabled={!this.isEnable('resume')}>
          恢复
        </Menu.Item>
      </Menu>
    )

    return (
      <Dropdown overlay={menu}>
        <Button size="small" data-testid="dropdown">
          操作
          <Icon
            data-testid="icon"
            type={current.loading ? 'loading' : 'down'}
          />
        </Button>
      </Dropdown>
    )
  }
}
