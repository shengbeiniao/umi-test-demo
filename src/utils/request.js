import { extend } from 'umi-request'

const request = extend({
  prefix:
    process.env.NODE_ENV === 'test'
      ? `http://172.20.11.243:3000/mock/72/api/cloud/zstack/v1`
      : '/api/cloud/zstack/v1'
})

request.use(async (ctx, next) => {
  let { req } = ctx
  let { options } = req
  let { params } = options
  if (process.env.NODE_ENV === 'test') {
    params.test = true
  }
  await next()
})

export default request
