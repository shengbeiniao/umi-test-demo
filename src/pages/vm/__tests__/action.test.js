import {
  render,
  waitForDomChange,
  fireEvent,
  wait,
  waitForElement
} from '@testing-library/react'
import { delay, initDva } from 'utils/jest'
import Action from '../Action'

let store

const renderPage = async current => {
  let vmModel = require('models/vm.js').default
  vmModel.state = {
    ...vmModel.state,
    list: [
      {
        uuid: '1',
        state: '运行中'
      },
      {
        uuid: '2',
        state: '已停止'
      },
      {
        uuid: '3',
        state: '已暂停'
      }
    ]
  }
  const { app, Page } = initDva([vmModel], <Action current={current} />)
  const root = render(<Page />)
  const { getByTestId,debug } = root
  fireEvent.mouseEnter(getByTestId('dropdown'))
  await waitForDomChange({ container: document.body })
  store = app._store
  return root
}

test('render menu/运行中', async () => {
  const current = {
    uuid: '1',
    state: '运行中'
  }
  const { getByTestId } = await renderPage(current)
  expect(getByTestId('menu')).toMatchSnapshot()
})

test('render menu/已停止', async () => {
  const current = {
    uuid: '2',
    state: '已停止'
  }
  const { getByTestId } = await renderPage(current)
  expect(getByTestId('menu')).toMatchSnapshot()
})

test('render menu/已暂停', async () => {
  const current = {
    uuid: '3',
    state: '已暂停'
  }
  const { getByTestId } = await renderPage(current)
  expect(getByTestId('menu')).toMatchSnapshot()
})

test('action/stop', async () => {
  const current = {
    uuid: '1',
    state: '运行中'
  }
  const { getByTestId, debug } = await renderPage(current)
  fireEvent.click(getByTestId('stop'))
  await delay(1000)
  const newCurrent = store
    .getState()
    ['vm'].list.find(item => item.uuid === current.uuid)
  expect(newCurrent.state).toEqual('已停止')
})

test('action/start', async () => {
  const current = {
    uuid: '2',
    state: '已停止'
  }
  const { getByTestId, debug } = await renderPage(current)
  fireEvent.click(getByTestId('start'))
  await delay(1000)
  const newCurrent = store
    .getState()
    ['vm'].list.find(item => item.uuid === current.uuid)
  expect(newCurrent.state).toEqual('运行中')
})

test('action/pause', async () => {
  const current = {
    uuid: '1',
    state: '运行中'
  }
  const { getByTestId, debug } = await renderPage(current)
  fireEvent.click(getByTestId('pause'))
  await delay(1000)
  const newCurrent = store
    .getState()
    ['vm'].list.find(item => item.uuid === current.uuid)
  expect(newCurrent.state).toEqual('已暂停')
})

test('action/resume', async () => {
  const current = {
    uuid: '3',
    state: '已暂停'
  }
  const { getByTestId, debug } = await renderPage(current)
  fireEvent.click(getByTestId('resume'))
  await delay(1000)
  const newCurrent = store
    .getState()
    ['vm'].list.find(item => item.uuid === current.uuid)
  expect(newCurrent.state).toEqual('运行中')
})
