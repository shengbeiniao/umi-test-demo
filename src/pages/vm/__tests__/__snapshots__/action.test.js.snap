// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render menu/已停止 1`] = `
<ul
  class="ant-dropdown-menu ant-dropdown-menu-light ant-dropdown-menu-root ant-dropdown-menu-vertical"
  data-testid="menu"
  role="menu"
  tabindex="0"
>
  <li
    aria-disabled="false"
    class="ant-dropdown-menu-item"
    data-testid="start"
    role="menuitem"
  >
    启动
  </li>
  <li
    aria-disabled="true"
    class="ant-dropdown-menu-item ant-dropdown-menu-item-disabled"
    data-testid="stop"
    role="menuitem"
  >
    停止
  </li>
  <li
    aria-disabled="true"
    class="ant-dropdown-menu-item ant-dropdown-menu-item-disabled"
    data-testid="pause"
    role="menuitem"
  >
    暂停
  </li>
  <li
    aria-disabled="true"
    class="ant-dropdown-menu-item ant-dropdown-menu-item-disabled"
    data-testid="resume"
    role="menuitem"
  >
    恢复
  </li>
</ul>
`;

exports[`render menu/已暂停 1`] = `
<ul
  class="ant-dropdown-menu ant-dropdown-menu-light ant-dropdown-menu-root ant-dropdown-menu-vertical"
  data-testid="menu"
  role="menu"
  tabindex="0"
>
  <li
    aria-disabled="true"
    class="ant-dropdown-menu-item ant-dropdown-menu-item-disabled"
    data-testid="start"
    role="menuitem"
  >
    启动
  </li>
  <li
    aria-disabled="false"
    class="ant-dropdown-menu-item"
    data-testid="stop"
    role="menuitem"
  >
    停止
  </li>
  <li
    aria-disabled="true"
    class="ant-dropdown-menu-item ant-dropdown-menu-item-disabled"
    data-testid="pause"
    role="menuitem"
  >
    暂停
  </li>
  <li
    aria-disabled="false"
    class="ant-dropdown-menu-item"
    data-testid="resume"
    role="menuitem"
  >
    恢复
  </li>
</ul>
`;

exports[`render menu/运行中 1`] = `
<ul
  class="ant-dropdown-menu ant-dropdown-menu-light ant-dropdown-menu-root ant-dropdown-menu-vertical"
  data-testid="menu"
  role="menu"
  tabindex="0"
>
  <li
    aria-disabled="true"
    class="ant-dropdown-menu-item ant-dropdown-menu-item-disabled"
    data-testid="start"
    role="menuitem"
  >
    启动
  </li>
  <li
    aria-disabled="false"
    class="ant-dropdown-menu-item"
    data-testid="stop"
    role="menuitem"
  >
    停止
  </li>
  <li
    aria-disabled="false"
    class="ant-dropdown-menu-item"
    data-testid="pause"
    role="menuitem"
  >
    暂停
  </li>
  <li
    aria-disabled="true"
    class="ant-dropdown-menu-item ant-dropdown-menu-item-disabled"
    data-testid="resume"
    role="menuitem"
  >
    恢复
  </li>
</ul>
`;
